crates/ra_assists/src/handlers/number_representation.rs

```diff
@@ -0,0 +1,603 @@
+        check_assist_not_applicable_with_id(
+            separate_number_literal,
+            SEPARATE_BINARY_NIBBLES_ID,
+            r#"fn f() { let x = <|>0b1010;}"#,
+        );
+    }
+
+    #[test]
+    fn separate_number_literal_binary_nibbles_too_small_separator_not_applicable() {
+        check_assist_not_applicable_with_id(
+            separate_number_literal,
+            SEPARATE_BINARY_NIBBLES_ID,
+            r#"fn f() { let x = <|>0b1_01_0;}"#,
+        );
+    }
+
+    #[test]
+    fn separate_number_literal_binary_nibbles() {
+        check_assist_with_id(
+            separate_number_literal,
+            SEPARATE_BINARY_NIBBLES_ID,
+            r#"fn f() { let x = <|>0b0010101000101010; }"#,
+            r#"fn f() { let x = <|>0b0010_1010_0010_1010; }"#,
+        )
+    }
+
+    #[test]
+    fn separate_number_literal_binary_nibbles_badly_split() {
+        check_assist_with_id(
+            separate_number_literal,
+            SEPARATE_BINARY_NIBBLES_ID,
+            r#"fn f() { let x = <|>0b001_0101_000_101_010; }"#,
+            r#"fn f() { let x = <|>0b0010_1010_0010_1010; }"#,
+        )
+    }
+
+    // ---
+
+    #[test]
+    fn separate_number_literal_binary_bytes_target() {
+        check_assist_target_with_id(
+            separate_number_literal,
+            SEPARATE_BINARY_BYTES_ID,
+            r#"fn f() { let x = <|>0b0010101000101010; }"#,
+            r#"0b0010101000101010"#,
+        );
+    }
+
+    #[test]
+    fn separate_number_literal_binary_bytes_already_split_not_applicable() {
+        check_assist_not_applicable_with_id(
+            separate_number_literal,
+            SEPARATE_BINARY_BYTES_ID,
+            r#"fn f() { let x = <|>0b00101010_00101010; <|>}"#,
+        );
+    }
+
+    #[test]
+    fn separate_number_literal_binary_bytes_too_small_not_applicable() {
+        check_assist_not_applicable_with_id(
+            separate_number_literal,
+            SEPARATE_BINARY_BYTES_ID,
+            r#"fn f() { let x = <|>0b00101010;}"#,
+        );
+    }
+
+    #[test]
+    fn separate_number_literal_binary_bytes_too_small_separator_not_applicable() {
+        check_assist_not_applicable_with_id(
+            separate_number_literal,
+            SEPARATE_BINARY_BYTES_ID,
+            r#"fn f() { let x = <|>0b0_0_101_01_0;}"#,
+        );
+    }
+
+    #[test]
+    fn separate_number_literal_binary_bytes() {
+        check_assist_with_id(
+            separate_number_literal,
+            SEPARATE_BINARY_BYTES_ID,
+            r#"fn f() { let x = <|>0b0010101000101010; }"#,
+            r#"fn f() { let x = <|>0b00101010_00101010; }"#,
+        )
+    }
+
+    #[test]
+    fn separate_number_literal_binary_bytes_badly_split() {
+        check_assist_with_id(
+            separate_number_literal,
+            SEPARATE_BINARY_BYTES_ID,
+            r#"fn f() { let x = <|>0b001_0101_000_101_010; }"#,
+            r#"fn f() { let x = <|>0b00101010_00101010; }"#,
+        )
+    }
```

Comment:
That's a lot of tests :)

Reply:
I'm not sure all the combinations are useful in the end they are essentially the same tests applied to all supported representations, can remove some if needed :)

Reply:
yeah, I think it might make sense to remove some of the boilerplate here with a table driven test:

```
for (before, after) in [("0b001_0101_000_101_010", "0b00101010_00101010") ... ] 
```